### PR TITLE
No checkout flag is specified only for server side clone.

### PR DESCRIPTION
**Pull request by varshavaradarajan**
#2152

### CODE REVIEW

File: common/src/com/thoughtworks/go/config/materials/git/GitMaterial.java [hunk 1]
```diff
@@ -235,7 +235,13 @@ public void afterCompletion(int status) {
                 });
             }
             int cloneDepth = shallowClone ? preferredCloneDepth : Integer.MAX_VALUE;
-            int returnValue = gitCommand.cloneFrom(outputStreamConsumer, url.forCommandline(), cloneDepth);
+            int returnValue;
+            if(executionContext.isServer()) {
+                returnValue = gitCommand.serverSideClone(outputStreamConsumer, url.forCommandline());
```
Comment (tomzo) [inline common/src/com/thoughtworks/go/config/materials/git/GitMaterial.java]: Hi @varshavaradarajan 
I am wondering if there is any way that checkout on server can be still executed? There is `ConfigMaterialUpdater` which relies on `MaterialPoller.checkout` to actually make a clean checkout of git repository. The path starts around [here](https://github.com/gocd/gocd/blob/master/server/src/com/thoughtworks/go/server/materials/ConfigMaterialUpdater.java#L114) and IIRC it eventually calls `updateTo` which it seems to me, now would never execute a full checkout if running on server side.

Comment (varshavaradarajan) [inline common/src/com/thoughtworks/go/config/materials/git/GitMaterial.java]: Hi @tomzo, 

Previously, the ```--no-checkout``` flag was there on both the server and the agent side. When ```updateTo``` was called from the agent side, the following are executed:
```
git fetch origin --prune
GitCommand.resetWorkingDir()
```
The above would update the working copy to the given revision, which is the behaviour of ```git checkout``` 

So, when you call ```MaterialPoller.checkout``` from ```ConfigMaterialUpdater```, the working copy on the server side does get updated to the revision that you specify.

Comment (zabil) [inline common/src/com/thoughtworks/go/config/materials/git/GitMaterial.java]: > I am wondering if there is any way that checkout on server can be still executed?

Yes, If the material is a configuration repository. So this feature will only work for materials that don't store configuration.

File: common/src/com/thoughtworks/go/domain/materials/git/GitCommand.java [hunk 2]
```diff
@@ -54,17 +54,22 @@ public GitCommand(String materialFingerprint, File workingDir, String branch, bo
         this.environment = environment;
     }
 
-    public int cloneFrom(ProcessOutputStreamConsumer outputStreamConsumer, String url) {
-        return cloneFrom(outputStreamConsumer, url, Integer.MAX_VALUE);
+    public int serverSideClone(ProcessOutputStreamConsumer outputStreamConsumer, String url) {
```
Comment (zabil) [inline common/src/com/thoughtworks/go/domain/materials/git/GitCommand.java]: call this method ```cloneWithNoCheckout```

**Pull request merged**